tools/fs/safe-watcher.ts

```diff
@@ -1,476 +1,364 @@
-    return {
-      close() {
-        entry.release(callback);
-      }
-    } as SafeWatcher;
+export async function closeAllWatchers() {
+  for (const root of Array.from(watchRoots)) {
+    await safeUnsubscribeSub(root);
   }
-);
+}
 
-const fireNames = {
-  [nsfw.actions.CREATED]: 'change',
-  [nsfw.actions.MODIFIED]: 'change',
-  [nsfw.actions.DELETED]: 'delete'
+function hasPriority(absPath: string) {
+  // If we're not prioritizing changed files, then all files have
+  // priority, which means they should be watched with native file
+  // watchers if the platform supports them. If we are prioritizing
+  // changed files, then only changed files get priority.
+  return PRIORITIZE_CHANGED
+      ? changedPaths.has(absPath)
+      : true;
 }
 
-export function addWatchRoot(absPath: string) {
-  if (watchRoots.has(absPath) || watcherLibrary !== 'nsfw' || !watcherEnabled) {
-    return;
+// Determines the polling interval to be used for the fs.watchFile-based
+// safety net that works on all platforms and file systems.
+function getPollingInterval(absPath: string): number {
+  if (hasPriority(absPath)) {
+    // Regardless of whether we have a native file watcher and it works
+    // correctly on this file system, poll prioritized files (that is,
+    // files that have been changed at least once) at a higher frequency
+    // (every 500ms by default).
+    return NO_WATCHER_POLLING_INTERVAL;
   }
 
-  watchRoots.add(absPath);
-
-  // If there already is a watcher for a parent directory, there is no need
-  // to create this watcher.
-  for (const path of watchRoots) {
-    let relativePath = pathRelative(path, absPath);
-    if (
-      path !== absPath &&
-      !relativePath.startsWith('..') &&
-      !relativePath.startsWith('/')
-    ) {
-      return;
-    }
+  if (watcherEnabled || PRIORITIZE_CHANGED) {
+    // As long as native file watching is enabled (even if it doesn't
+    // work correctly) and the developer hasn't explicitly opted out of
+    // the file watching priority system, poll unchanged files at a
+    // lower frequency (every 5000ms by default).
+    return DEFAULT_POLLING_INTERVAL;
   }
 
-  // TODO: check if there are any existing watchers that are children of this
-  // watcher and stop them
-
-  nsfw(
-    convertToOSPath(absPath),
-    (events) => {
-      events.forEach(event => {
-        if(event.action === nsfw.actions.RENAMED) {
-          let oldPath = nativeJoin(event.directory, event.oldFile);
-          let oldEntry = entries[toPosixPath(oldPath)];
-          if (oldEntry) {
-            oldEntry._fire('rename');
-          }
+  // If native file watching is disabled and the developer has
+  // explicitly opted out of the priority system, poll everything at the
+  // higher frequency (every 500ms by default). Note that this leads to
+  // higher idle CPU usage, so the developer may want to adjust the
+  // METEOR_WATCH_POLLING_INTERVAL_MS environment variable.
+  return NO_WATCHER_POLLING_INTERVAL;
+}
```

Comment:
Both (`DEFAULT_POLLING_INTERVAL` and `NO_WATCHER_POLLING_INTERVAL`) are configured to be focused on performance, right?

What would be those settings if I'm on a machine with fewer resources(_maybe_ we could have high-perf mode and low one)?

Reply:
This code was already in the previous watcher, I just moved it around.

Polling is a fallback, only enabled manually with `METEOR_WATCH_FORCE_POLLING`, and the polling interval can be adjusted. Parcel’s native watching remains the default with my change. I initially planned to remove polling, but kept it in case someone needs it. It’s not expected to be used often, but if needed, the interval can be configured via `METEOR_WATCH_POLLING_INTERVAL_MS` as they can already do.